### PR TITLE
[Enhancement] add fine grained range predicate rule 

### PR DESCRIPTION
**Pull request by packy92**
## Why I'm doing:
For certain scenarios, customers require querying aggregated results over a large time span. Direct querying performance cannot meet the requirements. It's necessary to utilize precomputed results of aggregation mv on a daily, monthly, and yearly dimension to accelerate the queries. To match the pattern of the materialized view, we need to divide the time span into multiple segments.

## What I'm doing:
Introducing the RewriteAtMostOnceTask which rewrites the whole tree by TopDown way. If a node had been rewritten, we don't need process the children and just finish the task.

Introducing the FineGrainedRangePredicateRule. This rule operates as follows:
- Rewriting is only permitted for patterns that satisfy SPJG, and the rewriting process requires enabling `enable_fine_grained_range_predicate`
- Divides the time span into the following segments.
image

- Based on the segmented time span, splits the original predicates into fine-grained mutually exclusive intervals.
- Replaces the original predicates with the new fine-grained predicates, creating new sub-OptExpressions through duplication.
- After unioning the sub-OptExpressions, rewrites the original aggCall using the output columns of the union.
The transformation process is like:
```
/*
 *         agg                            new_agg
 *          ｜                               |
 *        filter         =>               union
 *          ｜                        /      |      \
 *        input                     agg1   agg2    ...
 *                                   |       |      |
 *                                filter1 filter2  ...
 *
 */
```

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSetType.java [hunk 1]
```diff
@@ -38,5 +38,7 @@ public enum RuleSetType {
     ALL_MV_REWRITE,
     PRUNE_EMPTY_OPERATOR,
     SHORT_CIRCUIT_SET,
-    NUM_RULE_SET
+    NUM_RULE_SET,
+
+    FINE_GRAINED_RANGE_PREDICATE
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSetType.java]: fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+                            .addChildren(Pattern.create(OperatorType.PATTERN_LEAF))));
+
+    public static final FineGrainedRangePredicateRule PROJECTION_INSTANCE = new FineGrainedRangePredicateRule(
+            RuleType.TF_FINE_GRAINED_RANGE_PREDICATE,
+            Pattern.create(OperatorType.LOGICAL_AGGR)
+                    .addChildren(Pattern.create(OperatorType.LOGICAL_PROJECT)
+                            .addChildren(Pattern.create(OperatorType.LOGICAL_FILTER)
+                                    .addChildren(Pattern.create(OperatorType.PATTERN_LEAF)))));
+
+    private static final Set<String> SUPPORTED_FUNC = ImmutableSet.of(FunctionSet.COUNT, FunctionSet.SUM,
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java]: Yes, I add some other fucntions that we can support and the comment to explain why these fucntions.

**Review by ABingHuang — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteAtMostOnceTask.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        super(context, root, rules, true);
+    }
+
+    @Override
+    protected void rewrite(OptExpression parent, int childIndex, OptExpression root) {
+
+        if (change > 0) {
+            return;
+        }
+        super.rewrite(parent, childIndex, root);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteAtMostOnceTask.java:40]: the task look same as `onlyOnce` task?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteAtMostOnceTask.java:40]: emmm, i get it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+        }
+
+        List<ScalarOperator> otherPredicates = Lists.newArrayList();
+        Map<ColumnRefOperator, List<BinaryPredicateOperator>> columnToRange = Maps.newHashMap();
+        extractRangePredicate(filterOperator.getPredicate(), otherPredicates, columnToRange);
+        if (columnToRange.isEmpty()) {
+            return Lists.newArrayList();
+        }
+
+        // only handle [,]/(,]/[/)/[]
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java:138]:
```suggestion
        // only handle [,]/(,]/[,)/[]
```

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java:138]: can't support `()` ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitDatePredicateRule.java [hunk 5]
```diff
@@ -1,294 +0,0 @@
...
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-
-import static com.starrocks.catalog.Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF;
-
-// For query like select sum() from xxx where  k1 >= "2022-01-02" and "2024-05-06" >= k1 group by xxx
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/SplitDatePredicateRule.java:53]: save the comment?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java (hunk 4 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/FineGrainedRangePredicateRule.java]: support. I modify the comment.

**Review by Seaven — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Pull request merged**